src/mds/CDir.cc

```diff
@@ -2160,10 +2161,166 @@ class C_IO_Dir_Committed : public CDirIOContext {
     dir->_committed(r, version);
   }
   void print(ostream& out) const override {
-    out << "dirfrag_commit(" << dir->dirfrag() << ")";
+    out << "dirfrag_committed(" << dir->dirfrag() << ")";
   }
 };
 
+class C_IO_Dir_Commit_Ops : public Context {
+public:
+  C_IO_Dir_Commit_Ops(CDir *d, version_t v, int pr, bufferlist &bl,
+                      vector<dentry_key_t> &rm, vector<CDir::dentry_commit_items> &s,
+                      vector<string> &_stale)
+    : dir(d), version(v), op_prio(pr) {
+    dfts.swap(bl);
+    to_remove.swap(rm);
+    to_set.swap(s);
+    _stale_items.swap(_stale);
+  }
+
+  void finish(int r) override {
+    dir->_omap_commit_ops(r, op_prio, version, dfts, to_remove, _stale_items,
+                          to_set);
+  }
+
+private:
+  CDir *dir;
+  version_t version;
+  int op_prio;
+  bufferlist dfts;
+  vector<dentry_key_t> to_remove;
+  vector<CDir::dentry_commit_items> to_set;
+  vector<string> _stale_items;
+};
+
+void CDir::_omap_commit_ops(int r, int op_prio, version_t version, bufferlist &dfts,
+                            vector<dentry_key_t> &to_remove, vector<string> &_stales,
+                            vector<dentry_commit_items> &to_set)
+{
+  dout(10) << __func__ << dendl;
+
+  if (r < 0) {
+    mdcache->mds->handle_write_error_with_lock(r);
+    return;
+  }
+
+  C_GatherBuilder gather(g_ceph_context,
+                         new C_OnFinisher(new C_IO_Dir_Committed(this, version),
+			 mdcache->mds->finisher));
+
+  SnapContext snapc;
+  object_t oid = get_ondisk_object();
+  object_locator_t oloc(mdcache->mds->mdsmap->get_metadata_pool());
+
+  map<string, bufferlist> _set;
+  set<string> _rm;
+
+  unsigned max_write_size = mdcache->max_dir_commit_size;
+  unsigned write_size = 0;
+
+  char *dftstr = dfts.c_str();
+
+  auto commit_one = [&](bool header=false) {
+    ObjectOperation op;
+
+    // don't create new dirfrag blindly
+    if (!is_new())
+      op.stat(nullptr, nullptr, nullptr);
```

Comment:
can't use is_new in this context

Reply:
Because this is not under the mds_lock and the `new_item` need to proect by it, right ?
We should add one flag before this and then check the flag here.